### ISSUE TITLE
JDK patch upgrade campaign (April 2024)

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#4058]
### Service(s)

cert.ci.jenkins.io, ci.jenkins.io, infra.ci.jenkins.io, release.ci.jenkins.io, trusted.ci.jenkins.io, weekly.ci.jenkins.io

### Summary

As per https://www.java.com/releases/, all JDKs we are using (8, 11, 17 and 21) received a patch upgrade the 16 April 2024.
This issue tracks the steps required to upgrade all JDKs across the platform for our developers.

The scope is for JDK tools, Jenkins agent JDKs and Jenkins controllers JDKs. Note that we accept a patch-difference between controller and agent JDks.

New JDK versions:

- 21.0.3, Available on adoptium: https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21.0.3%2B9
- 17.0.11, Available on adoptium: https://github.com/adoptium/temurin17-binaries/releases/tag/jdk-17.0.11%2B9
- 11.0.23, Available on adoptium: https://github.com/adoptium/temurin11-binaries/releases/tag/jdk-11.0.23%2B9
- 8u412, Available on adoptium: https://github.com/adoptium/temurin8-binaries/releases/tag/jdk8u412-b08

Elements to be updated:

- Jenkins controller images:
  - [x] Weekly 2.455 (23 April 2024) use the new JDKs 21.0.3 and 17.0.11(for infra.ci and weekly.ci controllers) 
  - [ ] Weekly 2.456 (30 April 2024) is expected to use new JDK 11.0.23 for containers
  - [ ] Next LTS 2.452.1 (15 May 2024) is expected to use the new JDKs (for all the other controllers)
- Jenkins agent images: 
  - [x] [Packer Image, aka. "all-in-one" images](https://github.com/jenkins-infra/packer-images) for Linux VM agents, Windows VM agents and most of the Linux container agents.
      - JDK8: https://github.com/jenkins-infra/packer-images/pull/1139
      - JDK11: https://github.com/jenkins-infra/packer-images/pull/1142
      - JDK17: https://github.com/jenkins-infra/packer-images/pull/1137
      - JDK21: https://github.com/jenkins-infra/packer-images/pull/1139
      - Release: https://github.com/jenkins-infra/packer-images/releases/tag/1.65.0
      - Deployment: https://github.com/jenkins-infra/jenkins-infra/pull/3396
  - [x] [Windows Container agent](https://github.com/jenkins-infra/docker-inbound-agents)
      - JDK8: https://github.com/jenkins-infra/docker-inbound-agents/pull/161
      - JDK11: https://github.com/jenkins-infra/docker-inbound-agents/pull/158
      - JDK17: https://github.com/jenkins-infra/docker-inbound-agents/pull/159
      - JDK21: https://github.com/jenkins-infra/docker-inbound-agents/pull/160
      - Deployment: https://github.com/jenkins-infra/jenkins-infra/pull/3399
  - [ ] [Web-builder aka. "docker-builder"](https://github.com/jenkins-infra/docker-builder)
  - [ ] [Packaging aka. "docker-packaging"](https://github.com/jenkins-infra/docker-packaging)
- JDK tools:
  - Puppet-managed controllers:
    - [x] ci.jenkins.io
    - [x] trusted.ci.jenkins.io
    - [x] cert.ci.jenkins.io
    - PRs:
      - JDK8: https://github.com/jenkins-infra/jenkins-infra/pull/3386
      - JDK11: https://github.com/jenkins-infra/jenkins-infra/pull/3391
      - JDK17: https://github.com/jenkins-infra/jenkins-infra/pull/3392
      - JDK21: https://github.com/jenkins-infra/jenkins-infra/pull/3387
  - Kubernetes-managed controllers:
    - [x] infra.ci.jenkins.io
      - JDK8: https://github.com/jenkins-infra/kubernetes-management/pull/5148
      - JDK11: https://github.com/jenkins-infra/kubernetes-management/pull/5149
      - JDK17: https://github.com/jenkins-infra/kubernetes-management/pull/5173
      - JDK21: https://github.com/jenkins-infra/kubernetes-management/pull/5174

**Comment by MarkEWaite** [comment]
I have submitted a pull request that should allow updatecli to run for the controller container images:

* https://github.com/jenkinsci/docker/pull/1852

**Comment by MarkEWaite** [comment]
Confirmed that JDK 21.0.3 and JDK 17.0.11 are included in the containers for the weekly release 2.455 and that JDK 11.0.22 containers in the weekly release 2.455.

**Comment by dduportal** [comment]
Update:

- All JDK tools on Puppet-based controllers (ci.jenkins.io/trusted.ci.jenkins.io/cert.ci.jenkins.io) have been upgraded with no problem using the automatic detection with `updatecli`
- All JDK installations have been upgraded on the "all in one" templates for agents (Linux VM, Windows VM and Linux containers) and the new image has been deployed
- JDK tools for infra.ci.jenkins.io (kubernetes-manegement) have only been upgraded partially: only JDK8 and JDK11. As underlined in https://github.com/jenkins-infra/kubernetes-management/pull/4998, we have to fix the updatecli manifest to ensure JDK17 and JDK21 are upgraded => WiP
- Windows container image are WiP:
  - JDK8, JDK11 and JDK17 upgrades had been detected by the `updatecli` automation but are blocked by https://github.com/jenkins-infra/helpdesk/issues/4048 => WiP
  - JDK21 upgrade had NOT been detected by the automation => WiP to add /fix the automation

**Comment by dduportal** [comment]
Update on the Windows Container agents:

- https://github.com/jenkins-infra/helpdesk/issues/4048 is now fixed allowing to work on the JDK upgrades
- JDK upgrades are not available yet as Docker images. https://github.com/jenkins-infra/docker-inbound-agents/pull/154 opened to switch from parent Docker image to using the Temurin installed like we did/are doing for the official Jenkins images => WiP

**Comment by dduportal** [comment]
- Update of the JDK17 of the (parent image of docker/packaging and docker/webbuilder) jenkinsci/agent Docker image is on hold: https://github.com/jenkinsci/docker-agent/pull/796 because of https://github.com/jenkinsci/docker-agent/pull/786
- Jenkins LTS is next week

**Comment by dduportal** [comment]
- Docker agents have been updated with latest JDK:
  - https://github.com/jenkinsci/docker-agent/releases/tag/3206.vb_15dcf73f6a_9-11
  - Deployed to ci.jenkins.io with https://github.com/jenkins-infra/jenkins-infra/pull/3411
- Docker packaging has been updated:
  - https://github.com/jenkins-infra/docker-packaging/releases/tag/4.0.65
  - https://github.com/jenkins-infra/release/pull/537
- Docker webbuilder has been updated:
  - https://github.com/jenkins-infra/docker-builder/releases/tag/3.1.49
  - https://github.com/jenkins-infra/jenkins-infra/pull/3412
